File: front/lib/models/person.dart
/// Represents a family member with health data.
class Person {
  /// Optional: set when loaded from API.
  final String? id;
  final String name;
  final String phone;
  final DateTime dob;
  final String gender;
  final String bloodType;
  final String rhFactor;
  final double height; // meters
  final double weight; // kg
  final double sugarLevel; // g/L
  final int systolicBP;
  final int diastolicBP;
  final int heartRate;
  final List<String> allergies;
  final List<String> chronicDiseases;
  final bool vaccinesUpToDate;

  const Person({
    this.id,
    required this.name,
    required this.phone,
    required this.dob,
    required this.gender,
    required this.bloodType,
    required this.rhFactor,
    required this.height,
    required this.weight,
    required this.sugarLevel,
    required this.systolicBP,
    required this.diastolicBP,
    required this.heartRate,
    this.allergies = const [],
    this.chronicDiseases = const [],
    this.vaccinesUpToDate = true,
  });

  /// From API JSON (camelCase).
  factory Person.fromJson(Map<String, dynamic> json) {
    return Person(
      id: json['id'] as String?,
      name: json['name'] as String? ?? '',
      phone: json['phone'] as String? ?? '',
      dob: DateTime.tryParse(json['dob'] as String? ?? '') ?? DateTime.now(),
      gender: json['gender'] as String? ?? '',
      bloodType: json['bloodType'] as String? ?? '',
      rhFactor: json['rhFactor'] as String? ?? '',
      height: (json['height'] as num?)?.toDouble() ?? 0,
      weight: (json['weight'] as num?)?.toDouble() ?? 0,
      sugarLevel: (json['sugarLevel'] as num?)?.toDouble() ?? 0,
      systolicBP: json['systolicBP'] as int? ?? 0,
      diastolicBP: json['diastolicBP'] as int? ?? 0,
      heartRate: json['heartRate'] as int? ?? 0,
      allergies: (json['allergies'] as List<dynamic>?)?.cast<String>() ?? [],
      chronicDiseases: (json['chronicDiseases'] as List<dynamic>?)?.cast<String>() ?? [],
      vaccinesUpToDate: json['vaccinesUpToDate'] as bool? ?? true,
    );
  }

  Map<String, dynamic> toJson() => {
        if (id != null) 'id': id,
        'name': name,
        'phone': phone,
        'dob': dob.toIso8601String(),
        'gender': gender,
        'bloodType': bloodType,
        'rhFactor': rhFactor,
        'height': height,
        'weight': weight,
        'sugarLevel': sugarLevel,
        'systolicBP': systolicBP,
        'diastolicBP': diastolicBP,
        'heartRate': heartRate,
        'allergies': allergies,
        'chronicDiseases': chronicDiseases,
        'vaccinesUpToDate': vaccinesUpToDate,
      };

  /// BMI = weight / height^2 (kg/m²)
  double get bmi {
    if (height <= 0) return 0;
    return weight / (height * height);
  }

  /// Age from date of birth.
  int get age {
    final now = DateTime.now();
    int a = now.year - dob.year;
    if (now.month < dob.month || (now.month == dob.month && now.day < dob.day)) {
      a--;
    }
    return a;
  }

  Person copyWith({
    String? id,
    String? name,
    String? phone,
    DateTime? dob,
    String? gender,
    String? bloodType,
    String? rhFactor,
    double? height,
    double? weight,
    double? sugarLevel,
    int? systolicBP,
    int? diastolicBP,
    int? heartRate,
    List<String>? allergies,
    List<String>? chronicDiseases,
    bool? vaccinesUpToDate,
  }) {
    return Person(
      id: id ?? this.id,
      name: name ?? this.name,
      phone: phone ?? this.phone,
      dob: dob ?? this.dob,
      gender: gender ?? this.gender,
      bloodType: bloodType ?? this.bloodType,
      rhFactor: rhFactor ?? this.rhFactor,
      height: height ?? this.height,
      weight: weight ?? this.weight,
      sugarLevel: sugarLevel ?? this.sugarLevel,
      systolicBP: systolicBP ?? this.systolicBP,
      diastolicBP: diastolicBP ?? this.diastolicBP,
      heartRate: heartRate ?? this.heartRate,
      allergies: allergies ?? this.allergies,
      chronicDiseases: chronicDiseases ?? this.chronicDiseases,
      vaccinesUpToDate: vaccinesUpToDate ?? this.vaccinesUpToDate,
    );
  }
}

File: back/app/models/person.py
"""Person model (family member with health data)."""
from datetime import datetime
from enum import Enum
from typing import List, Optional

from pydantic import BaseModel, ConfigDict, Field


class PersonRole(str, Enum):
    FATHER = "father"
    MOTHER = "mother"
    CHILD = "child"


# Shared config: Python uses snake_case, JSON/Flutter receives camelCase aliases
_SHARED = ConfigDict(populate_by_name=True, serialize_by_alias=True)


class PersonBase(BaseModel):
    model_config = _SHARED

    name: str
    phone: str = ""
    dob: datetime
    gender: str
    blood_type: str = Field(alias="bloodType")
    rh_factor: str = Field(alias="rhFactor")
    height: float = Field(gt=0, le=3.0)                            # metres
    weight: float = Field(gt=0, le=500.0)                          # kg
    sugar_level: float = Field(alias="sugarLevel", ge=0.0, le=30.0)   # g/L
    systolic_bp: int = Field(alias="systolicBP", ge=50, le=300)       # mmHg
    diastolic_bp: int = Field(alias="diastolicBP", ge=30, le=200)     # mmHg
    heart_rate: int = Field(alias="heartRate", ge=20, le=300)         # bpm
    allergies: List[str] = Field(default_factory=list)
    chronic_diseases: List[str] = Field(default_factory=list, alias="chronicDiseases")
    vaccines_up_to_date: bool = Field(True, alias="vaccinesUpToDate")


class PersonCreate(PersonBase):
    family_id: str
    role: PersonRole


class PersonUpdate(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    name: Optional[str] = None
    phone: Optional[str] = None
    dob: Optional[datetime] = None
    gender: Optional[str] = None
    blood_type: Optional[str] = None
    rh_factor: Optional[str] = None
    height: Optional[float] = None
    weight: Optional[float] = None
    sugar_level: Optional[float] = None
    systolic_bp: Optional[int] = None
    diastolic_bp: Optional[int] = None
    heart_rate: Optional[int] = None
    allergies: Optional[List[str]] = None
    chronic_diseases: Optional[List[str]] = None
    vaccines_up_to_date: Optional[bool] = None


class PersonInDB(PersonBase):
    id: str
    family_id: str
    role: PersonRole
    created_at: datetime
    updated_at: datetime


class Person(PersonBase):
    """Complete patient model — inherits all vitals from PersonBase.

    Python code uses snake_case (e.g. person.sugar_level).
    JSON serialization produces camelCase for Flutter (e.g. sugarLevel).
    """

    id: str
    family_id: str = ""
    role: str  # father | mother | child

    @classmethod
    def from_mongo_doc(cls, doc: dict) -> "Person":
        """Build Person from a MongoDB document (snake_case keys)."""
        return cls(
            id=str(doc["_id"]),
            name=doc["name"],
            phone=doc.get("phone", ""),
            dob=doc["dob"],
            gender=doc["gender"],
            blood_type=doc.get("blood_type", ""),
            rh_factor=doc.get("rh_factor", ""),
            height=doc.get("height", 1.7),
            weight=doc.get("weight", 70.0),
            sugar_level=doc.get("sugar_level", 1.0),
            systolic_bp=doc.get("systolic_bp", 120),
            diastolic_bp=doc.get("diastolic_bp", 80),
            heart_rate=doc.get("heart_rate", 70),
            allergies=doc.get("allergies", []),
            chronic_diseases=doc.get("chronic_diseases", []),
            vaccines_up_to_date=doc.get("vaccines_up_to_date", True),
            role=doc.get("role", "child"),
            family_id=str(doc.get("family_id", "")),
        )
